src/content/docs/en/guides/middleware.mdx

```diff
@@ -0,0 +1,168 @@
+   const redactedHtml = html.replace("PRIVATE INFO", "REDACTED");
+   
+   return new Response(redactedHtml, {
+     status: 200,
+     headers: response.headers
+   });
+}
+```
+For more usage examples, please see [the middleware RFC proposal](https://github.com/withastro/roadmap/blob/rfc/stage-3-middleware/proposals/0032-middleware-api.md#middleware-workflow-and-examples)
+
+### Chaining middleware
+
+Multiple middlewares can be joined, in a specified order, using [`sequence()`](#sequence):
+
+```js title="src/middleware.js"
+import {sequence} from "astro/middleware";
+
+function validation() {}
+function auth() {}
+
+export const onRequest = sequence(validation, auth);
+```
+
+```js
+import {sequence} from "astro/middleware";
+
+async function validation(_, next) {
+    console.log("validation request");
+    const response = await next();
+    console.log("validation response");
+    return response;
+}
+async function auth(_, next) {
+  console.log("auth request");
+  const response = await next();
+  console.log("auth response");
+  return response;
+
+}
+async function greeting(_, next) {
+  console.log("greeting request");
+  const response = await next();
+  console.log("greeting response");
+  return response;
+
+}
+
+export const onRequest = sequence(validation, auth, greeting);
+```
+
+This will result in the following console order:
+
+```sh
+validation request
+auth request
+greeting request
+greeting response
+auth response
+validation response
+```
+
+## API reference
+
+### `onRequest()`
+
+A required exported function from `src/middleware.js` that will be called before rendering every page or API route. It accepts two optional arguments: [context](#context) and [next()](#next-)
+
+### `context`
+
+An object that includes information to be made available to other middleware, API routes and `.astro` routes during the rendering process.
+
+This is an optional argument passed to `onRequest()` that may contain the [`locals`](#locals) object as well as any additional properties to be shared during rendering. For example, the `context` object may include cookies used in authentication.
+
+This is the same `context` object that is passed to API routes.
+
+### `next()`
+
+A function that intercepts (reads and modifies) the `Response` of a `Request` or calls the "next" middleware in the chain and returns a `Response`. For example, this function could modify the HTML body of a response.
+
+This is an optional argument of `onRequest()`; however, your middleware **must** either return a `Response` or call `next()`.
+
+
+### `locals`
+An object containing data from a response that can be manipulated inside middleware. 
+9
+This `locals` object is forwarded across the request handling process and is available as a property to APIContext and AstroGlobal. This allows data to be shared between middlewares, API routes, and .astro pages. This is useful for storing request-specific data, such as user data, across the rendering step.
```

Comment:
FYI: currently, we disabled the middleware functionality for API routes ([endpoints](https://docs.astro.build/en/core-concepts/endpoints/#static-file-endpoints)?) that use the following pattern:

```js
export function get({}) {
	return {
		body: "some text"
	}
}
```

I mean, those endpoints that only return an object that contains `body`. Astro will log a warning then that happens.